docs/configuration.md
# Configuration

## mode
- syntax: mode client | server
- default: -
- required: yes
- mode: -

Set the running mode of dperf. dperf can be run as a client or a server.

## daemon
- syntax: daemon
- default: -
- required: no
- mode: client, server

In daemon mode, dperf statistics are written to the log file ('/var/log/dperf/dperf-ctl-[server|client].log'), otherwise output to stdout.

## quiet
- syntax: quiet
- default: -
- required: no
- mode: client, server

Turn off output statistics per second.

## keepalive
- syntax: keepalive interval(timeout) [num]
- default: -
- required: no
- mode: client, server

Keepalive means that multiple requests and responses can be sent in one connection.

For dperf server:
- When 'keepalive' is not enabled, the dperf server closes the connection at the fastest speed by setting FIN in the response message. (Think about it, can a program calling the POSIX socket API do this?)
- After enabling 'keepalive', the dperf server will wait for the client FIN/RST or a period of time before closing the connection.
- The connection idle timeout of the dperf server is the retransmission timeout (about 8 seconds) plus 'timeout'.
- 'num' does not need to be configured on the dperf server.

For dperf client:
- dperf client sends a request every 'interval' time; when the 'interval' is 0, the client will initiate a new request immediately after receiving the UDP response.
- The dperf client closed the connection after sending 'num' requests.

'keepalive' usage scenarios:
- CC test: set a larger number of concurrent connections (cc 10m), a larger interval, such as 60s.
- tps test: set a larger packet_size (such as 1500) and a smaller interval (such as 1ms).
- TX PPS test: set flood, smaller packet_size (eg 64), smaller number of concurrent connections cc (eg 3000), smaller interval (eg 1ms or 10us).
- TX and RX PPS test: set a smaller packet_size (such as 64), a smaller number of concurrent connections cc (such as 3000), and a smaller interval (such as 1ms or 1us).

## pipeline
- syntax: pipeline num
- default: 0
- required: no
- mode: client

UDP Client initiates NUM requests at the same time in a connection, which can increase PPS.

Example:
- one-way UDP elephant flow: cc 1; pipeline 4; keepalive 1us; flood
- two-way UDP elephant flow: cc 1; pipeline 4; keepalive 0us

## cpu
- syntax: cpu n0 n1 n2-n3...
- default: -
- required: yes
- mode: client, server

Set which CPUs dperf runs on. dperf is a multithread program.
It starts a worker thread on each CPU, and each thread uses only 1 RX queue and 1 TX queue.
If you use multiple network interface ports, dperf will distribute the CPU equally according to the configuration order of the network interface port.

On a multi-NUMA system, it is necessary to note that the CPU needs to be on the same NUMA node as the network interface port, which is a requirement of DPDK.

Example:
- cpu 0-3 12-15
- cpu 0 1 2 3 4 5

## socket_mem
- syntax: socket_mem n0,n1,n2...
- default: -
- required: no
- mode: client, server

This is the DPDK "--socket" parameter, and dperf passes it to DPDK. Using'--socket_mem',
we can run a dperf client and a dperf server on the same host at the same time,
so that we can build a load test environment with one host.

Example:
- socket_mem  40960,0
- socket_mem  0,40960

Note: The unit of the parameter is MB.

Reference:
[Linux-specific EAL parameters](http://doc.dpdk.org/guides/linux_gsg/linux_eal_parameters.html#linux-specific-eal-parameters)

[Multi-process Support](http://doc.dpdk.org/guides/prog_guide/multi_proc_support.html)

## port
- syntax: port PCI|BOND IPAddress Gateway [GatewayMAC]
- default: -
- required: yes
- mode: client, server

Configure the network interface port used by dperf.
If you want to use multiple network interface ports, you only need to configure multiple 'port's.
As a DPDK program, dperf will take over these network interface ports from the operating system.
Before starting dperf, you need to use the DPDK script 'dpdk-devbind.py' for driver binding (except for Mellanox network interfaces).
- PCI: The PCI number of the network interface port, use 'dpdk-devbind.py -s' to get it from the system.
- BOND: The format is bondMode:Policy(PIC0,PCI1,...), Mode is [0-4], Policy is [0-2].
- IPAddress: This IP  Address is used to interconnect with the 'Gateway'.
- Gateway: Gateway IP address. dperf has no routing capability. It will send all packets to the gateway, except for ARP, NS, and ND.
- Gateway-MAC: the MAC address of the 'Gateway', which can be omitted.

Example:
- port bond4:2(0000:81:10.0,0000:81:10.1) 10.235.20.12 10.235.20.1 00:00:64:01:01:01
- port 0000:03:00.1 6.6.215.4 6.6.215.1

Reference:

[binding-and-unbinding-network-ports](http://doc.dpdk.org/guides/linux_gsg/linux_drivers.html#binding-and-unbinding-network-ports-to-from-the-kernel-modules)

[Link Bonding Poll Mode Driver Library](https://doc.dpdk.org/guides/prog_guide/link_bonding_poll_mode_drv_lib.html)

## duration
- syntax: duration Time
- default: duration 100s
- required: yes
- mode: client, server

Set the running time of dperf, after this time dperf will exit.

After dperf is started, there is a slow start phase, and the CPS will gradually increase.
There is also a short buffer time when exiting. You can also use the signal 'SIGINT' to make dperf exit gracefully immediately.

Example:
- duration 1.5d
- duration 2h
- duration 3.5m
- duration 100s
- duration 100

## cps
- syntax: cps Number
- default: -
- required: yes
- mode: client

This is the total target for the number of new connections per second for all worker threads.
dperf evenly distributes the total target to each worker thread, so it is recommended to set'cps' to an integer multiple of the number of worker threads.
In the slow start phase, CPS will gradually increase.

Example:
- cps 10m
- cps 1.5m
- cps 2k
- cps 100

## cc
- syntax: cc Number
- default: -
- required: no
- mode: client

Set the total target for the number of concurrent connections on the client side.
When 'cc' is set, the client will enable 'keepalive'.
When the target value is high, we need to increase the time interval between two requests to reduce network bandwidth usage.

Example:
- cc 100m
- cc 1.5m
- cc 2k
- cc 100

## flood
- syntax: flood
- default: -
- required: no
- mode: client

Support TCP/UDP protocol, no need for network card to support FDIR.
If flood is enabled, the dperf client only sends the first packet of the connection; for the TCP protocol, dperf sends a SYN packet.

## launch_num
- syntax: launch_num Number
- default: launch_num 4
- required: no
- mode: client

How many connections are initiated by the client at a time.
In case of packet loss, try to reduce the number to make the packet sending more smooth,
so as to avoid the packet loss of the receiver's network card caused by the surge of packets.

## client
- syntax: client IPAddrStart IPAddrNumber
- default: -
- required: yes
- mode: client, server

Set client IP range:
-'ipaddrstart': starting address
-'ipaddrnumber': total addresses 1-254

Dperf uses the last two bytes of the IP address (including IPv4 and IPv6) to identify a client address.
In the address range of a 'client', only the last byte is allowed to change.
In the client mode, the number of 'clients' needs to be consistent with the number of' ports' and correspond one by one.
Indicates what client address pool this network port uses as the source address of the connection.
In the server mode, there is no requirement for the number of 'clients', which can be more or less than' port '.
Indicates which clients the server only receives. The dperf server does not accept connection requests from unknown addresses.

Example:

    The interface address and the client address are in the same network segment
    port        0000:1b:00.0    6.6.1.2   6.6.1.1
    client      6.6.1.2         100

    The interface address and the client address are not in the same network segment
    port        0000:1b:00.0    6.6.10.2   6.6.10.1
    port        0000:1b:00.1    6.6.20.2   6.6.20.1
    client      6.6.30.2        100
    client      6.6.40.2        100

## server
- syntax: server IPAddrStart IPAddrNumber
- default: -
- required: yes
- mode: client, server

Set the listening IP range of the server. The number of 'port' must be the same as the number of 'server'.
By default, the number of server IP addresses is required to be equal to the number of CPUs, that is, one for each worker thread.
When 'rss' is enabled, the number of server IP addresses is allowed to be less than the number of CPUs.

## listen
- syntax: listen Port Number
- default: listen 80 1
- required: yes

Set the port ranges that the server listens to, and the client sends packets to these port ranges.

Note: dperf will allocate all sockets at startup. Please do not set a large port range.

## payload_random
- syntax: payload_random
- default: -
- required: no
- mode: client, server

Set the payload to random characters ('a'-'z'), the default is all 'a'.

## payload_size
- syntax: payload_size Number(>=1)
- default: -
- required: no
- mode: client, server

Set the size of the request and response, in bytes, not including L2, L3, L4 header sizes.
For tcp protocol, if payload_size is less than 70, dperf will be forced to 70, which is the minimum HTTP packet length.
If you want to set smaller data packets, use packet_size.

## packet_size
- syntax: packet_size Number(0-1514)
- default: -
- required: no
- mode: client, server

Sets the data packet size, including the Ethernet header, excluding the 4-byte FCS. Use packet_size to set the minimum and maximum packets.

## jumbo
- syntax: jumbo
- default: -
- required: no
- mode: client, server

Enable jumbo frames. After enabling jumbo frames, packet_size can be set to 9724.
Note: packet_size cannot exceed the MTU of the network environment.

## mss
- syntax: mss Number
- default: mss 1460
- required: no
- mode: client, server

Set tcp MSS option.

## protocol
- syntax: protocol tcp | udp | http
- default: protocol tcp
- required: no
- mode: client, server

TCP or UDP or HTTP protocol. Regardless of the TCP or UDP protocol, the dperf client sends an HTTP Request, and the dperf server sends an HTTP Response.
Currently, 'http' only support client mode, you could use it to test nginx.

## tx_burst
- syntax: tx_burst Number(1-1024)
- default: tx_burst 8
- required: no
- mode: client, server

Use the DPDK API to send packets, the maximum number of packets sent at a time.
A smaller value can make the packets sending smoother and avoid packet loss at the receiving side, but it increases the CPU consumption of dperf.

## wait
- syntax: wait Seconds
- default: wait 3
- required: no
- mode: client

The time the client waits after startup before entering the slow-start phase.

## slow_start
- syntax: slow_start Seconds(10-600)
- default: slow_start 30
- required: no
- mode: client

The client gradually increases the number of new connections per second during the slow start time.

## vxlan
- syntax: vxlan vni innerSMAC innerDMAC localVtepIPAddr Number remoteVtepIPAddr Number
- default: -
- required: no
- mode: client, server

One 'vxlan' can be set per 'port'. 'innerSMAC' is the source MAC address of the inner packet, and 'innerDMAC' is the destination address of the inner packet.
'localVtepIPAddr' is the starting address of local vtep. Each NIC queue needs a local vtep address for traffic classification.
'remoteVtepIPAddr' is the starting address of the remote vtep. Number is the number of vtep addresses.

## kni
- syntax: kni [ifName]
- default: -
- required: no
- mode: client, server

Enable the kni interface. We create a kni interface for each 'port'. Common interface Names are vEth/vnic, etc. The default name is dperf.
The IP address and route of the kni interface need to be manually configured, and can only be configured as the 'port' IP.

## tos
- syntax: tos Number(0x00-0xff or 0-255)
- default: 0
- required: no
- mode: client, server

Set the tos of the IPv4 header or the traffic class of the IPv6 header, which can be in hexadecimal or decimal.
Note: tos does not take effect on the packets sent by the kni interface.

## rss
- syntax: rss [l3/l3l4/auto] [mq_rx_none|mq_rx_rss]
- default: -
- required: no
- mode: client, server

Use network card RSS distribution. This switch needs to be turned on when the network card without FDIR feature needs to enable multi-queue/multi-threading.
- l3: Use the IP address-based symmetric hash algorithm to offload traffic, requiring the network card to support modifying the RSS configuration, which is the default option
- l3l4: Use the l3l4 symmetric hash algorithm to split traffic, requiring the network card to support modifying the RSS configuration
- auto: Use the default algorithm of the network card to split traffic, which can be specified as l3 or l3l4, requiring only one server IP to be configured for a port
- mq_rx_rss: Use the DPDK parameter 'RTE_ETH_MQ_RX_RSS' to enable the RSS feature of the network card, which is the default parameter
- mq_rx_none: Do not use the DPDK parameter 'RTE_ETH_MQ_RX_RSS' to configure the network card, some network cards do not allow configuration, and can only be used in auto mode

Note: 'vxlan' cannot be used simultaneously with 'rss' yet.

## tcp_rst
- syntax: tcp_rst Number[0-1]
- default: 1
- required: no
- mode: client, server

Set whether dperf replies rst to SYN packets requesting unopened TCP ports.

## change_dip
- syntax: change_dip IPAddress Step Number
- default: -
- required: no
- mode: client

When 'flood' is enabled, 'change_ip' can be used to modify the destination IP of sending packets.
'change_ip' can be configured with multiple lines and supports up to 65536 IP addresses. These IP addresses are evenly distributed to each thread.
When dperf sends each TCP/UDP message, it modifies the destination IP of the message; dperf obtains the IP from the address pool by polling.
'IPAddress' is the starting address, 'Step' is the interval for the next IP, and 'Number' is the total number of IPs.

Example:
- change_ip 192.168.1.1 64 100
- change_dip 2001:6:6:241::1 1 20

## http_host
- syntax: http_host String(1-127)
- default: http_host dperf
- required: no
- mode: client

Set the host header of the HTTP request.

## http_path
- syntax: http_path String(1-255)
- default: http_path /
- required: no
- mode: client

Set the path of the HTTP request.

## lport_range
- syntax: lport_range Number [Number]
- default: 1 65535
- required: no
- mode: client

Set the local port range for clients. It is recommended to set the local port range to [1000-65535] on Google Cloud.

## client_hop
- syntax: client_hop
- default: -
- required: no
- mode: client

client changes ip and port at the same time when create new connections.


CHANGELOG.md
# Changelog

## v1.5.0 - 2023-03-10
### Added

- udp elephant flow

    thanks sheva7777
- client_hop

    thanks hgkiller
- supported Mellanox MT27500 [ConnectX-3]

    thanks SdotChen
- article: Using dperf to test the performance of DPVS(zh)

    thanks sheva7777
- article: How to install dperf on ubuntu(zh)

    thanks digger-yu
- article: Using dperf to test 100Gbps bandwidth(zh)

    thanks CHRIS123540
- article: DPVS v1.9.2 Performance Tests by dperf

    thanks ywc689
- article: dperf FAQ(zh)

### Changed

- print error numbers with red color

### Fixed

- flood with rss l3l4

    thanks sheva7777
- use unlikeyly to predict condition

    thanks panzhengyu
- address conflict in configuration file

    thanks CHRIS123540
- docs/configuration: update payload_size

## v1.4.0 - 2022-12-14
### Added

- support dpdk-22.11

    thanks ykzj

- support LoongArch

    thanks choumin

- payload random

    thanks uname-v

### Fixed

- crash at tcp reply reset

- vlan check

    thanks owenstake
- crash by kni_broadcast()

    thanks sheva7777

## v1.3.0 - 2022-09-01
### Added

- 10us keepalive request interval

    thanks jiawen94
- vlan

    thanks hgkiller
- config client local port range for google cloud

    thanks MichaelZhangCN, vsv1020, wanggaoli
- article: using dperf to test network bandwidth

    thanks thunderZH963(Hua Zhang)

### Fixed

- in 'rss auto' mode, the number of UDP concurrent connections is incorrectly counted
- mq_rx_none: don't set RTE_ETH_MQ_RX_RSS
- compile warning

    thanks digger-yu
- double semicolons

    thanks yangwenrui

## v1.2.0 - 2022-06-01
### Added

- http client
- change_dip: before the packet is sent, a huge IP pool is used to change the dest IP

### Changed

- rss support l3/l3l4/auto
- optimize statistics

### Fixed

- FIX: kni only use interface IP
- FIX: large rx/tx descriptor number
- FIX: print more message for bad gateway error
- FIX: tcp closing state
- FIX: separate tcp and http

## v1.1.0 - 2022-04-13
### Added

- vxlan

    thanks epsilon-rhk
- kni

    thanks Ppapzs
- bond

    thanks reinight@gmail.com, Ppapzs
- tos

    thanks Ppapzs
- packet_size

    thanks Ppapzs
- jumbo

    thanks Ppapzs
- wait stage
- rss
- quiet: turn off output statistics
- rtt

### Changed

- tcp/udp flood and dose not require FDIR
- support DPDK-20.11, DPDK-22.03
- server close idle sockets
- tcp rst on or off
- allow server ip num to be less than cpu number
- simplified keepalive configuration
- update readme
- related articles

    thanks metonymical

### Fixed

- FIX: client ip range check error
- FIX: failed to start using virtio

    thanks celia240
- FIX: divide by zero
- FIX: tcp stack bugs

    thanks sheva7777
- FIX: segmentation fault when there are few huge pages

    thanks yunfei_james@126.com
- FIX: byte order conversion bug

    thanks crezov

- FIX: docs

    thanks leohotfn

## v1.0.0 - 2022-02-22
### Added

- Slow start time configurable
- Configure files
- Design files

### Changed

- Change readme

### Fixed

- Fix ip csum (Issue #28)
- Fix compile warnning (Issue #20)
    
    thanks 1225951453
- Fix segfault at vmxnet3 (Issue #16)
    
    thanks Danshanshu
- Fix separate the log files of server and client
    
    thanks RaisonNiu 


src/config.h
/*
 * Copyright (c) 2021 Baidu.com, Inc. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * Author: Jianzhang Peng (pengjianzhang@baidu.com)
 */

#ifndef __CONFIG_H
#define __CONFIG_H

#include "ip_range.h"
#include "ip_list.h"
#include "port.h"
#include "eth.h"

#include <stdint.h>
#include <stdbool.h>
#include <netinet/in.h>

#define RTE_ARG_LEN         64
#define CACHE_ALIGN_SIZE    64
#define TCP_WIN             (1460 * 40)
#define NETWORK_PORT_NUM    65536

#define PACKET_SIZE_MAX     1514
#define DEFAULT_CPS         1000
#define DEFAULT_INTERVAL    1       /* 1s */
#define DEFAULT_DURATION    60
#define DEFAULT_TTL         64
#define ND_TTL              255
#define DEFAULT_LAUNCH      4
#define DELAY_SEC           4
#define WAIT_DEFAULT        3
#define SLOW_START_DEFAULT  30
#define SLOW_START_MIN      10
#define SLOW_START_MAX      600
#define KEEPALIVE_REQ_NUM   32767  /* 15 bits */

#define JUMBO_FRAME_MAX_LEN 0x2600
#define JUMBO_PKT_SIZE_MAX  (JUMBO_FRAME_MAX_LEN - ETHER_CRC_LEN)
#define JUMBO_MTU           (JUMBO_PKT_SIZE_MAX - 14)
#define JUMBO_MBUF_SIZE     (1024 * 11)
#define MBUF_DATA_SIZE      (1024 * 10)

#define MSS_IPV4            (PACKET_SIZE_MAX - 14 - 20 - 20)
#define MSS_IPV6            (PACKET_SIZE_MAX - 14 - 40 - 20)
#define MSS_JUMBO_IPV4      (JUMBO_PKT_SIZE_MAX - 14 - 20 - 20)
#define MSS_JUMBO_IPV6      (JUMBO_PKT_SIZE_MAX - 14 - 40 - 20)

#define DEFAULT_WSCALE      13

#define LOG_DIR             "/var/log/dperf"

#define HTTP_HOST_MAX       128
#define HTTP_PATH_MAX       256

#define HTTP_HOST_DEFAULT   "dperf"
#define HTTP_PATH_DEFAULT   "/"

#define TCP_ACK_DELAY_MAX   1024

#define RSS_NONE            0
#define RSS_L3              1
#define RSS_L3L4            2
#define RSS_AUTO            3

#define KNI_NAMESIZE        10

#define VLAN_ID_MIN         1
#define VLAN_ID_MAX         4094

#define PIPELINE_MIN        0
#define PIPELINE_MAX        100
#define PIPELINE_DEFAULT    0

struct config {
    bool server;
    bool keepalive;
    bool ipv6;
    bool vxlan;
    bool kni;
    bool daemon;
    bool flood;
    bool jumbo;
    bool payload_random;
    bool client_hop;
    uint8_t rss;
    bool mq_rx_rss;
    uint8_t rss_auto;
    bool quiet;
    bool tcp_rst;
    bool http;
    bool stats_http;    /* payload size >= HTTP_DATA_MIN_SIZE */
    uint8_t tos;
    uint8_t pipeline;
    uint8_t tx_burst;
    uint8_t protocol;   /* TCP/UDP */
    uint16_t vlan_id;

    int ticks_per_sec;

    int lport_min;
    int lport_max;

    char kni_ifname[KNI_NAMESIZE];
    int af;

    uint64_t keepalive_request_interval_us;
    /* tsc */
    uint64_t keepalive_request_interval;
    int keepalive_request_num;

    char http_host[HTTP_HOST_MAX];
    char http_path[HTTP_PATH_MAX];

    int payload_size;
    int packet_size;
    int mss;

    int wait;
    int slow_start;
    uint32_t launch_num;
    int duration;
    int cps;        /* connection per seconds */
    int cc;         /* current connections */

    int cpu[THREAD_NUM_MAX];
    int cpu_num;

    /*
     * Running multiple instances on one host
     * dpdk rte_eal_init() --socket-meme
     * */
    char socket_mem[RTE_ARG_LEN];

    struct netif_port ports[NETIF_PORT_MAX];
    int port_num;

    struct vxlan vxlans[NETIF_PORT_MAX];
    int vxlan_num;

    int listen;
    int listen_num;

    struct ip_group client_ip_group;
    struct ip_group server_ip_group;
    struct ip_list  dip_list;
};

#define config_for_each_port(cfg, port) \
    for ((port) = &((cfg)->ports[0]); ((port) - &((cfg)->ports[0])) < (cfg)->port_num; port++)

extern struct config g_config;
int config_parse(int argc, char **argv, struct config *cfg);
uint32_t config_get_total_socket_num(struct config *cfg, int id);
struct netif_port *config_port_get(struct config *cfg, int thread_id, int *p_queue_id);
void config_set_tsc(struct config *cfg, uint64_t hz);
void config_set_payload(struct config *cfg, char *data, int len, int new_line);

#endif

